### ISSUE TITLE
Type inference issue with spread and generics

### ISSUE THREAD

**Issue by flovilmart** [microsoft/TypeScript#37934]
We notice the following issue described in [this](https://typescript-v2-497.ortam.now.sh/dev/bug-workbench/?#code/C4TwDgpgBAglC8UDeUDaBrAXFAzsATgJYB2A5gLrZ5FlQC+AUKJFAEILJpa4EkXbEArgFsARhHz0oDBgBMIAYwA2AQ3zQFAe2J4oK7DADccxavVQtO4FFHZWxhgHpHUYAEYohHJ1QAPKrxklDw0pFAAPlBCYhKG9AyWuu4cKAB06SoANFDpqaL0DonWwhKkEADyogBWisA4HogAPAAq2QCqAHwAFPpQrTbYbQCUCB1QXWkZ2bn5dEMOzq4ATJ7ecABkbF0A7gAWhAq7qzYqskPZOLuagkqyNtAofgGhwdR8EVEi4vhxdNJFyw4JXwZUqNQUdTcPWyoiGMmY0AAsqUIC12mNECgGFAcVxPMQoABRXzKQTyRroCAgTQAMz62Up1LpnWCzQw5GMf02SGxuNQSnxRJJSjJqMZtKgbQZVIlzQ6wTa-I5DC5yF5OP5gvFdK6zQ+w1ZSv1RpVxgBwNB1Vq3ia-U6PWw-VskqG2GRINRdoxYwmOSmfry9HmMkWwBWXnxNIk6juj38IT4r0CYUi0W+vwS2iSAGYgSiwdboTZgwCAG7Pd6pr4SDjAbOpFSGIA) playground.

# TL;DR;

The type inferred when merging two objects trough a generic function is always resolved to `T & U` which is inexact if `T` shares keys with `U`. 

The correct inferred type of `const mergeObjects = <A, B>(a: A, b: B) => ({ ...a, ...b });`

should be 

```
type Merge<T, U> = { 
    [k in Exclude<keyof T, keyof U>]: T[k];
} & {
    [l in Exclude<keyof U, keyof T>]: U[l];
} & {
    [l in keyof (T | U)]: T[l] | U[l] 
};
<A, B>(a: A, b: B) => Merge<A, B>
```

As shown in the playground, the correct type is inferred when using the spread operator with concrete types.

**Comment by RyanCavanaugh** [comment]
The proposed definition is much more complex yet still not fully accurate because it can't account for aliasing effects.

See discussion in #28234

**Comment by flovilmart** [comment]
Thanks for pointing it out.
